contract/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+use near_contract_standards::fungible_token::receiver::FungibleTokenReceiver;
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::collections::{TreeMap, UnorderedMap};
+use near_sdk::{
+    env, json_types::U128, near_bindgen, serde_json, AccountId, BorshStorageKey, Gas,
+    PanicOnDefault, PromiseOrValue, PromiseResult,
+};
+
+use crate::errors::*;
+use crate::ext_interfaces::*;
+use crate::types::*;
+
+mod errors;
+mod ext_interfaces;
+mod on_transfer;
+mod types;
+
+pub const ONE_YOCTO: u128 = 1;
+pub const FT_TRANSFER_TGAS: Gas = Gas(30_000_000_000_000);
+pub const RESERVE_TGAS: Gas = Gas(50_000_000_000_000);
+
+#[derive(BorshSerialize, BorshStorageKey)]
+pub enum GamesKeys {
+    Games,
+    GamesWithKey { game_id: AccountId },
+    FinishedGames,
+    FinishedGamesWithKey { game_id: AccountId },
+}
+
+#[near_bindgen]
+#[derive(PanicOnDefault, BorshDeserialize, BorshSerialize)]
+pub struct Games {
+    version: u8,
+    token_address: AccountId,
+    games: UnorderedMap<AccountId, TreeMap<GameId, Game>>,
+    finished_games: UnorderedMap<AccountId, TreeMap<GameId, Game>>,
```

Comment:
why do we duplicate the same structures? 
games and finished_games have the same params
We can update a game status.